### PR TITLE
[TS migration] Migrate useLocalize

### PR DESCRIPTION
**Pull request by MaciejSWM**
### Fixed Issues

$ https://github.com/Expensify/App/issues/24940
PROPOSAL: 

### Tests

- Click and interact with the whole app - useLocalize is used everywhere so any regression will be immediately visible
- Verify that no errors appear in the JS console

### Offline tests

- Same as `Tests`

### QA Steps

- Click and interact with the whole app - useLocalize is used everywhere so any regression will be immediately visible

Use the date picker to set the date of brith in personal details, verify it works as expected and selected values are correct.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/85628901/387146f5-b392-4ab8-822b-1e33385ce9bd

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/85628901/01143a2f-71a1-4303-81d6-8ae43eb28cf3

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/85628901/bf9039ec-231a-4e6b-b934-8f14c6f3af56

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/85628901/558f55a3-0e21-48d7-bb89-ba2b2f5e429f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/85628901/98d3a1fc-2f20-4004-8769-dbcb5532e576

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/85628901/26cfc282-81af-49a4-a2d4-0a5b4ffa5f71

</details>

### CODE REVIEW

File: src/components/LocaleContextProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+}
+const Provider = compose(
+    withCurrentUserPersonalDetails,
+    withOnyx({
+        preferredLocale: {
+            key: ONYXKEYS.NVP_PREFERRED_LOCALE,
+            selector: (preferredLocale) => preferredLocale,
+        },
+    }),
+)(LocaleContextProvider);
```
Comment (blazejkustra) [inline src/components/LocaleContextProvider.tsx:123]: Change to:
```suggestion
const Provider = compose(
    withCurrentUserPersonalDetails,
    withOnyx<LocaleContextProviderProps, LocaleContextProviderOnyxProps>({
        preferredLocale: {
            key: ONYXKEYS.NVP_PREFERRED_LOCALE,
            selector: (preferredLocale) => preferredLocale ?? CONST.LOCALES.DEFAULT,
        },
    }),
)(LocaleContextProvider);
```

File: src/components/LocaleContextProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+type LocaleContextProviderProps = {
+    /** The user's preferred locale e.g. 'en', 'es-ES' */
+    preferredLocale: string;
+
+    /** The current user's personalDetails */
+    currentUserPersonalDetails: CurrentUserPersonalDetails;
+
+    /** Actual content wrapped by this component */
+    children: React.ReactNode;
+};
```
Comment (blazejkustra) [inline src/components/LocaleContextProvider.tsx:32]: Split into two:
```suggestion
type LocaleContextProviderOnyxProps = {
    /** The user's preferred locale e.g. 'en', 'es-ES' */
    preferredLocale: string;
};

type LocaleContextProviderProps = LocaleContextProviderOnyxProps & {
    /** The current user's personalDetails */
    currentUserPersonalDetails: CurrentUserPersonalDetails;

    /** Actual content wrapped by this component */
    children: React.ReactNode;
};
```

File: src/components/LocaleContextProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+const LocaleContext = createContext<LocaleContextProps>({
+    translate: () => '',
+    numberFormat: () => '',
+    datetimeToRelative: () => '',
+    datetimeToCalendarTime: () => '',
+    updateLocale: () => '',
+    formatPhoneNumber: () => '',
+    toLocaleDigit: () => '',
+    fromLocaleDigit: () => '',
+    preferredLocale: '',
```
Comment (blazejkustra) [inline src/components/LocaleContextProvider.tsx:71]: Use `CONST.LOCALES.DEFAULT` as a default

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/components/LocaleContextProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+type LocaleContextProps = {
+    translate: Translate;
+    numberFormat: NumberFormat;
+    datetimeToRelative: DatetimeToRelative;
+    datetimeToCalendarTime: DatetimeToCalendarTime;
+    updateLocale: UpdateLocale;
+    formatPhoneNumber: FormatPhoneNumber;
+    toLocaleDigit: ToLocaleDigit;
+    fromLocaleDigit: FromLocaleDigit;
+    preferredLocale: string;
```
Comment (blazejkustra) [inline src/components/LocaleContextProvider.tsx:61]: @MaciejSWM Can you add comments to describe these props?
```ts
    /** Returns translated string for given locale and phrase */
    translate: type

    /** Formats number formatted according to locale and options */
    numberFormat: type

    /** Converts a datetime into a localized string representation that's relative to current moment in time */
    datetimeToRelative: type

    /** Formats a datetime to local date and time string */
    datetimeToCalendarTime: type

    /** Updates date-fns internal locale */
    updateLocale: type

    /** Returns a locally converted phone number for numbers from the same region
     * and an internationally converted phone number with the country code for numbers from other regions */
    formatPhoneNumber: type

    /** Gets the standard digit corresponding to a locale digit */
    fromLocaleDigit: type

    /** Gets the locale digit corresponding to a standard digit */
    toLocaleDigit: type
```

File: src/components/LocaleContextProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    fromLocaleDigit: () => '',
+    preferredLocale: CONST.LOCALES.DEFAULT,
+});
+
+function LocaleContextProvider({preferredLocale = CONST.LOCALES.DEFAULT, currentUserPersonalDetails = {}, children}: LocaleContextProviderProps) {
+    const selectedTimezone = useMemo(() => currentUserPersonalDetails?.timezone?.selected, [currentUserPersonalDetails]);
+
+    const translate = useMemo<Translate>(() => (phrase, variables) => Localize.translate(preferredLocale, phrase, variables), [preferredLocale]);
+
+    const numberFormat = useMemo<NumberFormat>(() => (number, options) => NumberFormatUtils.format(preferredLocale, number, options), [preferredLocale]);
```
Comment (BartoszGrajdek) [inline src/components/LocaleContextProvider.tsx:99]: I noticed that some of these util functions (lines 95-116) accept strings, this is something we haven't noticed before, so I'd recommend going through all of them and making sure that they are strict enough - in cases where we use timezones, make use of that new type, and where we use LOCALES make use of ValueOf<typeof CONST.LOCALES>

**Review by BartoszGrajdek — APPROVED** [review]

File: src/components/LocaleContextProvider.tsx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+type LocaleContextProviderProps = LocaleContextProviderOnyxProps & {
+    /** The current user's personalDetails */
+    currentUserPersonalDetails: CurrentUserPersonalDetails;
+
+    /** Actual content wrapped by this component */
+    children: React.ReactNode;
+};
+
+type Translate = (phrase: string, variables: Record<string, string>) => string;
```
Comment (fabioh8010) [inline src/components/LocaleContextProvider.tsx:38]: @MaciejSWM let's make this type smarter, here is my idea:

```suggestion
type Translate = <TKey extends TranslationPaths>(
    phrase: TKey,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    variables?: TranslateType<EnglishTranslation, TKey> extends (...args: any) => any ? Parameters<TranslateType<EnglishTranslation, TKey>>[0] : never,
) => string;
```

This type will allow us to only use the keys that exist and, when is a key that needs variables, we will have type inference too.

```ts
translate('addDebitCardPage.nameOnCard');

translate('chronos.oooEventSummaryFullDay', {
    summary: '',
    dayCount: 1,
    date: '',
});
```

Note: You will have to export `TranslateType` and `TranslationPaths` types from `src/languages/types.ts`.

cc @blazejkustra

File: src/components/LocaleContextProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+type SelectedTimezone = {
+    /** Value of the selected timezone */
+    selected: (typeof TIMEZONES)[number];
+};
+
+type CurrentUserPersonalDetails = {
+    /** Timezone of the current user */
+    timezone?: SelectedTimezone;
+};
```
Comment (fabioh8010) [inline src/components/LocaleContextProvider.tsx:23]:
```suggestion
type CurrentUserPersonalDetails = Pick<PersonalDetails, 'timezone'>;
```

Shouldn't this be easier?

File: src/components/LocaleContextProvider.tsx [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+const Provider = compose(
+    withCurrentUserPersonalDetails,
+    withOnyx<LocaleContextProviderProps, LocaleContextProviderOnyxProps>({
+        preferredLocale: {
+            key: ONYXKEYS.NVP_PREFERRED_LOCALE,
+            selector: (preferredLocale) => preferredLocale ?? CONST.LOCALES.DEFAULT,
+        },
+    }),
+)(LocaleContextProvider);
+
```
Comment (fabioh8010) [inline src/components/LocaleContextProvider.tsx:146]:
```suggestion

Provider.displayName = 'withOnyx(LocaleContextProvider)';

```

I think we should keep this.

**Comment by MaciejSWM** [comment]
Thanks @fabioh8010 ! I implemented your suggestions and it's ready for re-review

File: src/components/LocaleContextProvider.tsx (hunk 6 above)
Comment (blazejkustra) [reply src/components/LocaleContextProvider.tsx]: @fabioh8010 We implemented a similar approach in this [PR](https://github.com/Expensify/App/pull/29742/files) (src/libs/Localize/index.ts, `translate` function). FYI Bartosz is still working on it

Comment (fabioh8010) [reply src/components/LocaleContextProvider.tsx]: @MaciejSWM @blazejkustra The way it was implemented in [that PR](https://github.com/Expensify/App/pull/29742/files) works better, let's use it then!

**Comment by MaciejSWM** [comment]
TS failing. I'm looking for a fix

**Review by blazejkustra — APPROVED** [review]

**Comment by blazejkustra** [comment]
Tests are failing 😢

File: src/components/LocaleContextProvider.tsx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    formatPhoneNumber: () => '',
+    toLocaleDigit: () => '',
+    fromLocaleDigit: () => '',
+    preferredLocale: CONST.LOCALES.DEFAULT,
+});
+
+function LocaleContextProvider({preferredLocale = CONST.LOCALES.DEFAULT, currentUserPersonalDetails = {}, children}: LocaleContextProviderProps) {
+    const selectedTimezone = useMemo(() => currentUserPersonalDetails?.timezone?.selected, [currentUserPersonalDetails]);
+
+    const translate = useMemo<Translate>(() => (phrase, variables) => Localize.translate(preferredLocale, phrase, variables), [preferredLocale]);
```
Comment (blazejkustra) [inline src/components/LocaleContextProvider.tsx:91]:
```suggestion
    const translate = useMemo<Translate>(() => (phrase, ...variables) => Localize.translate(preferredLocale, phrase, ...variables), [preferredLocale]);
```

File: src/components/LocaleContextProvider.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    WithCurrentUserPersonalDetailsProps & {
+        /** Actual content wrapped by this component */
+        children: React.ReactNode;
+    };
+
+type PhraseParameters<T> = T extends (...args: infer A) => string ? A : never[];
+
+type Phrase<TKey extends TranslationPaths> = TranslationFlatObject[TKey] extends (...args: infer A) => unknown ? (...args: A) => string : string;
+
+type Translate = <TKey extends TranslationPaths>(phrase: TKey, variables?: PhraseParameters<Phrase<TKey>>) => string;
```
Comment (blazejkustra) [inline src/components/LocaleContextProvider.tsx:30]:
```suggestion
type Translate = <TKey extends TranslationPaths>(phraseKey: TKey, ...phraseParameters: PhraseParameters<Phrase<TKey>>) => string;
```

**Review by blazejkustra — COMMENTED** [review]
Actually one more change is needed + we may need to wait for Localize [PR](https://github.com/Expensify/App/pull/29742/files) to be merged first (without it, typecheck is failing).

**Comment by fabioh8010** [comment]
@blazejkustra @MaciejSWM please let me know when this one is ready to review! Until there, I will uncheck this one from the spreadsheet.

**Review by kubabutkiewicz — APPROVED** [review]
lgtm!

**Comment by blazejkustra** [comment]
@thesahindia We would appreciate if you could work on the reviewer checklist this week as this PR is blocking other migration PRs, thanks!

File: src/libs/LocaleDigitUtils.ts [hunk 11]
```diff
@@ -1,13 +1,17 @@
 import _ from 'lodash';
+import {ValueOf} from 'type-fest';
+import CONST from '@src/CONST';
 import * as NumberFormatUtils from './NumberFormatUtils';
 
+type Locale = ValueOf<typeof CONST.LOCALES>;
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/libs/LocaleDigitUtils.ts:6]: We decided to for now keep it in the files that uses these CONST types, there will be a follow up task that cleans libs/hooks/HOCs and we will definitely move such types to a different file.

File: src/components/LocaleContextProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+    const updateLocale = useMemo<UpdateLocale>(() => () => DateUtils.setLocale(locale), [locale]);
+
+    const formatPhoneNumber = useMemo<FormatPhoneNumber>(() => (phoneNumber) => LocalePhoneNumber.formatPhoneNumber(phoneNumber), []);
+
+    const toLocaleDigit = useMemo<ToLocaleDigit>(() => (digit) => LocaleDigitUtils.toLocaleDigit(locale, digit), [locale]);
+
+    const fromLocaleDigit = useMemo<FromLocaleDigit>(() => (localeDigit) => LocaleDigitUtils.fromLocaleDigit(locale, localeDigit), [locale]);
+
+    const contextValue = useMemo(
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/components/LocaleContextProvider.tsx:117]: Yep, good catch. I fixed it!

File: src/components/LocaleContextProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+type NumberFormat = (number: number, options: Intl.NumberFormatOptions) => string;
+
+type DatetimeToRelative = (datetime: string) => string;
+
+type DatetimeToCalendarTime = (datetime: string, includeTimezone: boolean, isLowercase: boolean) => string;
+
+type UpdateLocale = () => void;
+
+type FormatPhoneNumber = (phoneNumber: string) => string;
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/components/LocaleContextProvider.tsx:38]: All these types are used in `LocaleContextProps` type and their respective implementation in useMemos. I agree it looks rather clunky, I refactored the code a bit, wdyt? cc @BartoszGrajdek / @kubabutkiewicz

**Comment by blazejkustra** [comment]
Friendly buump @allroundexperts

**Comment by BartoszGrajdek** [comment]
@allroundexperts kind bump 😄

**Review by allroundexperts — APPROVED** [review]
Looks good!

**Review by mountiny — APPROVED** [review]
Changes are looking good to me, lets be on a look out for any date related regressions maybe also with the datepicker

**Pull request merged**